Added Dlink MIBS, memory polling and temp polling for some devices

hanserasmus opened the pull request:
DO NOT DELETE THIS TEXT

#### Please note

> Please read this information carefully. You can run `./scripts/pre-commit.php` to check your code before submitting.

- [x] Have you followed our [code guidelines?](http://docs.librenms.org/Developing/Code-Guidelines/)

#### Testers

If you would like to test this pull request then please run: `./scripts/github-apply <pr_id>`, i.e `./scripts/github-apply 5926`

I don't have all the devices and FW combinations that Dlink has to offer, so the tests were done on the following devices:

### System info like FW version, HW version and Serial number now displaying properly:
Sucessful:
DGS-3620-28SC FW 3.00.004
DGS-3420-28SC FW 3.01.R003
DGS-3627G FW 3.00.B47
DGS-3420-28TC FW 3.01.R003 & FW 1.50.018
DES-3528 FW 3.00.012
DGS-3120-24TC FW 3.10.012

Not successful:
DES-3526 (OID returns empty)

### Temperature info:
Successful: 
DGS-3620-28SC FW 3.00.004
DGS-3420-28TC FW 1.50.018
DES-3528 FW 3.00.012
DGS-3120-24TC FW 3.10.012

Not successful:
DES-3526 & DGS-3627G  (Will never support it, switch does not have environment capabilities)
All DGS-3420-28xx with FW 3.01.R003. Seems to be a FW issue which we will take up with the vendors. We expect fixes in the near future in the FW release.

Screenshots of how the data displays is attached as well. 
![temperature](https://user-images.githubusercontent.com/24249616/34872535-e30ccb24-f799-11e7-9b70-027a2e5f4c5c.png)

![device-details](https://user-images.githubusercontent.com/24249616/34872549-ef95b586-f799-11e7-95c4-701ec37c1320.png)

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/librenms/librenms?pullRequest=8076) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/librenms/librenms?pullRequest=8076) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/librenms/librenms?pullRequest=8076) it.</sub>

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/librenms/librenms/inspections/8834677b-8e02-425b-bde8-baa3de95f2b1) was created.

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/librenms/librenms/inspections/8834677b-8e02-425b-bde8-baa3de95f2b1) completed: **No new issues**

murrant commented:
Looks pretty good.

Can you add test data for these?

`./scripts/save-tests-data -c -m mempools,processors,sensors -h HOSTNAME`

(Use -v if you need multiple sets of data)

hanserasmus commented:
On how many devices should I do this? One of each listed as successfully polled in my previous post?

laf commented:
If you have the time to do that then yes that would be great. It ensures we don't break support for these devices in the future.

hanserasmus commented:
Will do that tomorrow morning and post it back here.

hanserasmus commented:
OK so I have created the files you were looking for. Here they are. I have modified the filenames to resemble the Model number + FW version of the device. Hope this helps. Anything else?

[SNMPrec-results.zip](https://github.com/librenms/librenms/files/1630752/SNMPrec-results.zip)

murrant commented:
@hanserasmus I glanced at the data and I don't see any of the oids that hold the data `.1.3.6.1.4.1.171.12.11.1.9.4.1.`.  Did you collect the data with your patch applied?
 
Also, if you install snmpsim, you can check the data for yourself. https://docs.librenms.org/Developing/os/Test-Units/

hanserasmus commented:
@murrant 
The method I used was:
1) Install LibreNMS
2) Change the files in the current LibreNMS install to resemble the submitted files in git.
3) Keep track of all those files, and copy them to my local machine.
4) On my local machine I cloned a forked repo of LibreNMS.
5) Copied all of the changed files over to my local machine with the forked repo, and added them to the repo as per instruction on the site.

So in short, my current running system, on which I ran the save-test-data scripts, have all the files I have submitted, and is therefor "patched". Running `./validate.php ` results in:

./validate.php 
====================================
Component | Version
--------- | -------
LibreNMS  | 1.35-97-g788c0bd
DB Schema | 229
PHP       | 7.1.11
MySQL     | 5.5.56-MariaDB
RRDTool   | 1.4.8
SNMP      | NET-SNMP 5.7.2
====================================

[OK]    Database connection successful
[OK]    Database schema correct
[WARN]  Your local git contains modified files, this could prevent automatic updates.
	[FIX] You can fix this with ./scripts/github-remove
	Modified Files:
	 includes/polling/os/dlink.inc.php

I am honest in saying I may be a bit out of my depth here. I am not sure why the OID's do not match. All I know is, since the changes I have made, a lot more data like HW version and FW version are displaying correctly, and the temperatures are graphing as expected (with values as expected and physically seen with spot checks on the devices). I know this may not be good enough for keeping the code up to standard, and I understand that, but at the moment I have no other explanation.
I am even struggling to understand the `-v` switch in the test data scripts, and not sure how to implement that and if it would even make any difference. 
Sorry again.

laf commented on includes/definitions/discovery/dlink.yaml:
```diff
@@ -0,0 +1,12 @@
+                    value: swTemperatureCurrent
+                    num_oid: .1.3.6.1.4.1.171.12.11.1.8.1.2.
+                    descr: Current Sys Temp
+                    index: '{{ $index }}'
```
This should be `index: 'swTemperatureCurrent.{{ $index }}'`

laf commented on includes/definitions/discovery/dlink.yaml:
```diff
@@ -0,0 +1,12 @@
+                    oid: swTemperatureTable
+                    value: swTemperatureCurrent
+                    num_oid: .1.3.6.1.4.1.171.12.11.1.8.1.2.
+                    descr: Current Sys Temp
```
If you get more than one result back here then the descr will not be unique so just change it to: `descr: 'Current Sys Temp {{ $index }}'`

laf commented on includes/discovery/mempools/dlink.inc.php:
```diff
@@ -0,0 +1,24 @@
+ *
+ * @package    LibreNMS
+ * @link       http://librenms.org
+ * @copyright  2017 Thomas GAGNIERE
```
It's 2018 :)

laf commented on includes/polling/mempools/dlink.inc.php:
```diff
@@ -0,0 +1,25 @@
+$memory_oid = '.1.3.6.1.4.1.171.12.1.1.9.1.4.1';
+$perc = snmp_get($device, $memory_oid, '-OvQ');
+if (is_numeric($perc)) {
+    $mempool['perc']  = $perc;
```
You can delete this line, it's calculated by the code.

laf commented:
You're doing really well with it :) I've made some inline comments for a few things that need to be changed.

For the test data if you have the device in LibreNMS then run: `./scripts/save-test-data.php -h ID` - Change ID for the device ID.

hanserasmus commented:
@laf Thank you for this clear and precise indications of what you needed. I have made all the changes to the code you have commented on. I have not submitted it yet, I have just "patched" my current install with all the changes, and then ran the tests against that.  I have done the snmpsim install as well now, so I am adding the json file for that result as well.

Only thing that happened was, after I have made the changes to the files, and ran the test script, my historic temperature data (and ONLY temperature data) in the graphs disappeared. It now just shows current temperature, and nan in the historic graphs. Don't know if I broke something, or if it is just coincidence. All other devices are still running as expected and graphing as expected (with a glance over). 

Thanks again for the help and the trouble, I appreciate it!
[TestData.zip](https://github.com/librenms/librenms/files/1641824/TestData.zip)

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/librenms/librenms/inspections/da2505a5-8e67-4460-9ef0-c4207383fc7f) was created.

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/librenms/librenms/inspections/32eca4c4-4aa2-43d7-8572-b759c25fe3af) was created.

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/librenms/librenms/inspections/9b8ab2e9-ec2d-49ac-8a69-aab257b13b75) was created.

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/librenms/librenms/inspections/da2505a5-8e67-4460-9ef0-c4207383fc7f) completed: **No new issues**

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/librenms/librenms/inspections/9b8ab2e9-ec2d-49ac-8a69-aab257b13b75) completed: **No new issues**

laf commented:
> Only thing that happened was, after I have made the changes to the files, and ran the test script, my historic temperature data (and ONLY temperature data) in the graphs disappeared for the device (ID 14) on which I ran the test. It now just shows current temperature, and "nan" in the historic graphs. Don't know if I broke something, or if it is just coincidence. All other devices are still running as expected and graphing as expected (with a glance over).

Is it still showing NaN?

hanserasmus commented:
@laf No it has been creating new historic data, so that is not a problem anymore.

laf commented:
@hanserasmus Ok are you happy if we merge this now or are you working on anymore bits?

hanserasmus commented:
I am happy at the moment. Will work on some extra bits as I go a long (but will still be a while). All the patches seem to work well, so I am happy. Thanks for the trouble @laf

murrant commented:
Thanks @hanserasmus.  Congrats :)

laf commented:
![image](https://user-images.githubusercontent.com/3941142/35198452-115d3668-fee7-11e7-9ae0-7ff92dfa05cb.png)

hanserasmus commented:
Hehe thanks you guys for putting up with me! :-)
